app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/PeerGroupHandshakeAsyncTask.java

```diff
@@ -0,0 +1,158 @@
+ * (no. of files & their names) with the receiver. Finally, the onPostExecute() of the sender
+ * initiates the file transfer through {@link SenderDeviceAsyncTask} on the sender and using
+ * {@link ReceiverDeviceAsyncTask} on the receiver.
+ */
+class PeerGroupHandshakeAsyncTask extends AsyncTask<Void, Void, InetAddress> {
+
+  private static final String TAG = "PeerGrpHndshakeAsyncTsk";
+  private final String HANDSHAKE_MESSAGE = "Request Kiwix File Sharing";
+
+  private WeakReference<LocalFileTransferActivity> weakReferenceToActivity;
+  private WifiP2pInfo groupInfo;
+
+  public PeerGroupHandshakeAsyncTask(LocalFileTransferActivity localFileTransferActivity, WifiP2pInfo groupInfo) {
+    this.weakReferenceToActivity = new WeakReference<>(localFileTransferActivity);
+    this.groupInfo = groupInfo;
+  }
+
+  @Override
+  protected InetAddress doInBackground(Void... voids) {
+    if(BuildConfig.DEBUG)  {
+      Log.d(TAG, "Handshake in progress");
+    }
+
+    if (groupInfo.groupFormed && groupInfo.isGroupOwner && !isCancelled()) {
+      try (ServerSocket serverSocket = new ServerSocket(PEER_HANDSHAKE_PORT)) {
+        serverSocket.setReuseAddress(true);
+        Socket server = serverSocket.accept();
+
+        ObjectInputStream objectInputStream = new ObjectInputStream(server.getInputStream());
+        Object object = objectInputStream.readObject();
+
+        // Verify that the peer trying to communicate is a kiwix app intending to transfer files
+        if (isKiwixHandshake(object) && !isCancelled()) {
+          if (BuildConfig.DEBUG) {
+            Log.d(TAG, "Client IP address: " + server.getInetAddress());
+          }
+          exchangeFileTransferMetadata(server.getOutputStream(), server.getInputStream());
+
+          return server.getInetAddress();
+        } else { // Selected device is not accepting wifi direct connections through the kiwix app
+          return null;
+        }
+      } catch (Exception ex) {
+        ex.printStackTrace();
+        return null;
+      }
+    } else if (groupInfo.groupFormed && !isCancelled()) { //&& !groupInfo.isGroupOwner
+      try (Socket client = new Socket()) {
+        client.setReuseAddress(true);
+        client.connect((new InetSocketAddress(groupInfo.groupOwnerAddress.getHostAddress(),
+            PEER_HANDSHAKE_PORT)), 15000);
+
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(client.getOutputStream());
+        objectOutputStream.writeObject(
+            HANDSHAKE_MESSAGE); // Send message for the peer device to verify
+
+        exchangeFileTransferMetadata(client.getOutputStream(), client.getInputStream());
+
+        return groupInfo.groupOwnerAddress;
+      } catch (Exception ex) {
+        ex.printStackTrace();
+        return null;
+      }
+    } else {
+      return null;
+    }
+  }
+
+  private boolean isKiwixHandshake(Object object) {
+    return (object.getClass().equals(String.class) && object.equals(HANDSHAKE_MESSAGE));
```

Comment:
Actually this can be easily simplified to 
```
HANDSHAKE_MESSAGE.equals(object)
```
The constant on the left hand side of the expression is generally better as it avoids you having to perform null checks and the like

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/58c936b4fec2f7589cac2249f324c1d74146667a#diff-b6bbaac42877486339bfdb7cc97c9aa4R103)